pilot/pkg/bootstrap/certcontroller.go

```diff
@@ -126,10 +93,94 @@ func (s *Server) initCertController(args *PilotArgs) error {
 	return nil
 }
 
-func pilotDNSCertDir() (string, error) {
-	userHomeDir, err := os.UserHomeDir()
+// initDNSCerts will create the certificates to be used by Istiod GRPC server and webhooks, signed by K8S server.
+// If the certificate creation fails - for example no support in K8S - returns an error.
+// Will use the mesh.yaml DiscoveryAddress to find the default expected address of the control plane,
+// with an environment variable allowing override.
+//
+// TODO: If the discovery address in mesh.yaml is set to port 15012 (XDS-with-DNS-certs) and the name
+// matches the k8s namespace, failure to start DNS server is a fatal error.
+func (s *Server) initDNSCerts() error {
+	if _, err := os.Stat(DNSCertDir + "/key.pem"); err == nil {
+		// Existing certificate mounted by user. Skip self-signed certificate generation.
+		// Use this with an existing CA - the expectation is that the cert will match the
+		// DNS name in DiscoveryAddress.
+		return nil
+	}
+
+	// discAddr configured in mesh config - this is what we'll inject into pods.
+	discAddr := s.mesh.DefaultConfig.DiscoveryAddress
+
+	// Explicit override - discovery address may point to a remote cluster.
+	// Istiod will be a local service doing only injection
+	if IstiodAddress.Get() != "" {
+		discAddr = IstiodAddress.Get()
```

Comment:
That's good, because we're planning to rename it :-)
Original plan was to have a new service and deployment named istiod and remove the rest - @howardjohn proposed instead to just merge it all in pilot and rename at the end.